### PR TITLE
Fix initialization of optimizers in DDP Strategy

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?
Redo of #11886
also fixes: #12317

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/strategies/sharded.py [hunk 1]
```diff
@@ -40,6 +40,23 @@ class DDPShardedStrategy(DDPStrategy):
...
+        # move the model to the correct device
+        self.model_to_device()
+
+        if self.sync_batchnorm:
+            self.model = self.configure_sync_batchnorm(self.model)
+
+        # skip wrapping the model if we are not fitting as no gradients need to be exchanged
+        trainer_fn = trainer.state.fn
+        if trainer_fn == TrainerFn.FITTING:
+            self.configure_ddp()
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:58]: Sharded inheriting from DDP makes this unnecessarily complicated considering they have different behaviors. This is why issues like https://github.com/PyTorchLightning/pytorch-lightning/issues/11863 are important

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/strategies/sharded.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/strategies/sharded.py:58]: We are missing a call to super().setup() here. Which means we won't setup optimizers there which is fine I guess because we do that in configure_ddp for this strategy. 

Just want to make sure you are aware that the missing super.setup()  means we won't call `accelerator.setup()` and `self.setup_precision_plugin() `

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/strategies/sharded.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/strategies/sharded.py:58]: @awaelchli - I am holding off on this until https://github.com/PyTorchLightning/pytorch-lightning/pull/11985 is reviewed & merged. then i'll redo the optimizer initialization, precision plugin setup, and accelerator setup

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/strategies/sharded.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/strategies/sharded.py:59]: I cannot call `super().setup()` because that will all DDPStrategy.setup() which does something completely different: it wraps the model in DistributedDataParallel, not ShardedDataParallel!

Comment (ananthsub) [reply pytorch_lightning/strategies/sharded.py]: @awaelchli I don't even think we can rely on `super().setup()` inside of DDP since we should call the accelerator.setup(trainer) *before* we call `model_to_device` to handle any accelerator validation. 
https://github.com/PyTorchLightning/pytorch-lightning/blob/b5fe056765ce71b237c641a1b159c346e5e8bd33/pytorch_lightning/accelerators/gpu.py#L46-L50

at this point i really dislike how `setup` is packaged on the base strategy. i'd much prefer calling the individual methods at different points rather than try to dance around them.

File: pytorch_lightning/strategies/ddp.py [hunk 2]
```diff
@@ -152,6 +154,17 @@ def setup(self, trainer: "pl.Trainer") -> None:
         if trainer_fn == TrainerFn.FITTING:
             self.configure_ddp()
 
+        # set up optimizers after the wrapped module has been moved to the device
+        self.setup_optimizers(trainer)
+        self.setup_precision_plugin()
+        optimizers_to_device(self.optimizers, self.root_device)
```
Comment (ananthsub) [inline pytorch_lightning/strategies/ddp.py]: `setup_optimizers` returns early if the trainer isn't fitting or tuning.
but `setup_precision_plugin` looks up the model, optimizer, and lr scheduler configs.

are all precision plugins built to handle optimizers/LR schedulers not being available outside of fitting/tuning? 

https://github.com/PyTorchLightning/pytorch-lightning/blob/b5fe056765ce71b237c641a1b159c346e5e8bd33/pytorch_lightning/strategies/strategy.py#L146-L152

File: pytorch_lightning/strategies/fully_sharded.py [hunk 3]
```diff
@@ -135,16 +135,15 @@ def setup_distributed(self) -> None:
 
     def setup(self, trainer: "pl.Trainer") -> None:
         self.accelerator.setup(trainer)
-        self.setup_optimizers(trainer)
-        self.setup_precision_plugin()
-        optimizers_to_device(self.optimizers, self.root_device)
```
Comment (ananthsub) [inline pytorch_lightning/strategies/fully_sharded.py:138]: we were calling `configure_optimizers` 3 times in this one method 😞 
i can split these out as separate PRs if desired

File: pytorch_lightning/strategies/ddp_spawn.py [hunk 4]
```diff
@@ -119,10 +121,15 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
 
         # skip wrapping the model if we are not fitting as no gradients need to be exchanged
-        trainer_fn = self.lightning_module.trainer.state.fn
+        trainer_fn = trainer.state.fn
         if trainer_fn == TrainerFn.FITTING:
             self.configure_ddp()
 
+        # set up optimizers after the wrapped module has been moved to the device
+        self.setup_optimizers(trainer)
+        optimizers_to_device(self.optimizers, self.root_device)
```
Comment (ananthsub) [inline pytorch_lightning/strategies/ddp_spawn.py:130]: wondering if `self.setup_optimizers` should include `optimizers_to_device` by default, so users only ever need to call `setup_optimizers`

File: pytorch_lightning/strategies/ddp.py [hunk 5]
```diff
@@ -135,12 +136,13 @@ def setup_environment(self) -> None:
         super().setup_environment()
 
     def setup(self, trainer: "pl.Trainer") -> None:
-        super().setup(trainer)
         # share ddp pids to all processes
         self._rank_0_will_call_children_scripts = self.broadcast(self._rank_0_will_call_children_scripts)
         if self._should_run_deadlock_detection():
             self._share_information_to_prevent_deadlock()
 
+        self.accelerator.setup(trainer)
```
Comment (tchaton) [inline pytorch_lightning/strategies/ddp.py:144]: Why this change? This doesn't seem related to this PR.

File: pytorch_lightning/strategies/ddp.py [hunk 6]
```diff
@@ -152,6 +154,17 @@ def setup(self, trainer: "pl.Trainer") -> None:
         if trainer_fn == TrainerFn.FITTING:
             self.configure_ddp()
 
+        # set up optimizers after the wrapped module has been moved to the device
+        self.setup_optimizers(trainer)
```
Comment (tchaton) [inline pytorch_lightning/strategies/ddp.py:158]: Let's try to make the code more readable.

Let's call super().setup(trainer) and override `setup_optimizers` to add the post_localSGD within its own method.

```
super().setup(trainer)

def setup_optimizers(self):
    super().setup_optimizers()
    self._reinit_optimizers_with_post_localSGD() # contain the torch version checks
```

File: pytorch_lightning/strategies/ddp.py [hunk 7]
```diff
@@ -257,6 +264,7 @@ def _reinit_optimizers_with_post_localSGD(self, warmup_steps: int):
             optimizers[x] = post_localSGD_optimizer
             del optimizer
         self.optimizers = optimizers
+        optimizers_to_device(self.optimizers, self.root_device)
```
Comment (tchaton) [inline pytorch_lightning/strategies/ddp.py:267]: This shouldn't be needed with the proposed refactor above as this would be called in the parent setup class?

File: pytorch_lightning/strategies/ddp_spawn.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/strategies/ddp_spawn.py:130]: Is `PostLocalSGDState` not supported for DDPSpawn? Why do we have a feature divergence there ?

Comment (tchaton) [reply pytorch_lightning/strategies/ddp_spawn.py:130]: This is already in the parent setup method. I would prefer to rely as much as possible on the parent methods from the base `Strategy`.

Maybe the base class should be:

```py
    def setup(self, trainer: "pl.Trainer") -> None:
        """Setup plugins for the trainer fit and creates optimizers.

        Args:
            trainer: the trainer instance
        """
        self.accelerator.setup(trainer)
        self.model_to_device()
        self.configure_model()
        self.setup_optimizers(trainer)
        self.setup_precision_plugin()
        optimizers_to_device(self.optimizers, self.root_device)
```

File: pytorch_lightning/strategies/fully_sharded.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/strategies/fully_sharded.py:138]: Would be better.

File: pytorch_lightning/strategies/ddp_spawn.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/strategies/ddp_spawn.py:130]: @awaelchli @carmocca @justusschock

File: pytorch_lightning/strategies/ddp.py (hunk 6 above)
Comment (awaelchli) [inline pytorch_lightning/strategies/ddp.py:158]: by moving the `setup_optimizers()` after the configure_ddp(), it will now affect the BaguaStrategy. 
Bagua runs checks on the optimizer in configuer_ddp:

```py
        if self._bagua_algorithm == "qadam":
            self._check_qadam_optimizer()
```

We would need to move it also after the setup_optimizer() call was made.

Comment (ananthsub) [reply pytorch_lightning/strategies/ddp.py:158]: great catch @awaelchli ! i forgot that bagua also extending DDP :/

File: pytorch_lightning/strategies/ddp_spawn.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/strategies/ddp_spawn.py:130]: as @awaelchli  this doesn't work even now. Bagua expects the optimizers to be initialized before wrapping the model.

File: pytorch_lightning/strategies/tpu_spawn.py [hunk 8]
```diff
@@ -140,7 +137,8 @@ def setup(self, trainer: "pl.Trainer") -> None:
             set_shared_parameters(self.model.module, shared_params)
 
         self.setup_optimizers(trainer)
-        self.precision_plugin.connect(self.model, None, None)
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/strategies/tpu_spawn.py:144]: @kaushikb11 could you take a look?

**Comment by rohitgr7** [comment]
hey @ananthsub!
is this blocked on something or ready to go?

**Comment by rohitgr7** [comment]
assigned myself. Will complete this soon.

File: pytorch_lightning/strategies/ddp_spawn.py [hunk 9]
```diff
@@ -186,6 +189,10 @@ def configure_ddp(self) -> None:
         self.model = self._setup_model(LightningDistributedModule(self.model))
         self._register_ddp_hooks()
 
+        # set up optimizers after the wrapped module has been moved to the device
+        self.setup_optimizers(self.lightning_module.trainer)
+        optimizers_to_device(self.optimizers, self.root_device)
```
Comment (awaelchli) [inline pytorch_lightning/strategies/ddp_spawn.py:194]: why not place it in the setup method as with the DDPStrategy? These two will need to collapse anyway in the future.

File: pytorch_lightning/strategies/sharded.py [hunk 10]
```diff
@@ -40,16 +41,41 @@ class DDPShardedStrategy(DDPStrategy):
     strategy_name = "ddp_sharded"
     _REDUCE_BUFFER_SIZE_DEFAULT: int = 2**23  # 8M
 
-    def configure_ddp(self) -> None:
-        trainer = self.lightning_module.trainer
-        if "reduce_buffer_size" not in self._ddp_kwargs:
-            # For multi-node training, enabling bucketing will improve performance.
-            self._ddp_kwargs["reduce_buffer_size"] = self._REDUCE_BUFFER_SIZE_DEFAULT if self.num_nodes > 1 else 0
+    def setup(self, trainer: "pl.Trainer") -> None:
```
Comment (awaelchli) [inline pytorch_lightning/strategies/sharded.py:44]: isn't this the same as the DDP implementation? The only difference I see is in `configure_ddp`

Comment (rohitgr7) [reply pytorch_lightning/strategies/sharded.py:44]: no.. there's a configuration for `PostLocalSGDState` as well.

**Comment by austinmw** [comment]
Any updates on this PR?

**Review by carmocca — APPROVED** [review]
